### PR TITLE
SE - Non-nullable value types: Add support  for Enum constraint

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6105

`where T: Enum` should learn `ObjectConstraint.NotNull` for all symbols of `T`.

~Blocked by #6949~

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs [hunk 1]
```diff
@@ -57,9 +57,18 @@ internal sealed class FieldReference : SimpleProcessor<IFieldReferenceOperationW
...
+        var state = context.State;
+        if (fieldReference.WrappedOperation.TrackedSymbol() is IFieldSymbol fieldSymbol)
+        {
+            if (context.State[fieldSymbol] is { } value)
+            {
+                state = state.SetOperationValue(context.Operation, value);
+            }
+            if (fieldSymbol.Type.IsNonNullableValueType() || fieldSymbol.Type.IsEnum())
+            {
+                state = state.SetSymbolConstraint(fieldSymbol, ObjectConstraint.NotNull);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:69]: This set the NotNull constraint for e.g. value types in `this` fields, static fields, or for enum values.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs [hunk 2]
```diff
@@ -94,6 +94,27 @@ public override ProgramState[] PostProcess(SymbolicContext context)
                     };
                 }
             }
+            else if (context.Operation.Instance.AsObjectCreation() is { } objectCreation
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs:97]: When I added the constraint, the lock-released tests started to fail and while I was tracking this down, I realized, that the logic for this rule was not right. I don't know what we want to do with this change as it is unrelated to the sprint.

This change actually does the same as what was done in line 84 above, just for a different overload.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Mutex.cs [hunk 3]
```diff
@@ -23,8 +23,6 @@ public void Noncompliant(Foo foo)
             var m3 = Mutex.OpenExisting("x");
             m3.WaitOne(); // Noncompliant
 
-            var m4 = new Mutex();
-
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Mutex.cs:27]: This needed to be removed because the graph was exploding too much.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Extensions/IOperationExtensions.cs [hunk 4]
```diff
@@ -30,6 +30,7 @@ internal static class IOperationExtensions
                 OperationKindEx.LocalReference => ILocalReferenceOperationWrapper.FromOperation(operation).Local,
                 OperationKindEx.ParameterReference => IParameterReferenceOperationWrapper.FromOperation(operation).Parameter,
                 OperationKindEx.Argument => IArgumentOperationWrapper.FromOperation(operation).Value.TrackedSymbol(),
+                OperationKindEx.DeclarationExpression => IDeclarationExpressionOperationWrapper.FromOperation(operation).Expression.TrackedSymbol(),
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Extensions/IOperationExtensions.cs:33]: I don't understand this change. Is it intentional? If I delete it, I can't find a failing UT.
```suggestion
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs [hunk 5]
```diff
@@ -57,9 +57,18 @@ internal sealed class FieldReference : SimpleProcessor<IFieldReferenceOperationW
 
     protected override ProgramState Process(SymbolicContext context, IFieldReferenceOperationWrapper fieldReference)
     {
-        var state = fieldReference.WrappedOperation.TrackedSymbol() is { } fieldSymbol && context.State[fieldSymbol] is { } value
-            ? context.State.SetOperationValue(context.Operation, value)
-            : context.State;
+        var state = context.State;
+        if (fieldReference.WrappedOperation.TrackedSymbol() is IFieldSymbol fieldSymbol)
+        {
+            if (context.State[fieldSymbol] is { } value)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:63]: For consistency - when extracted, only the extracted state should be used. In case we add some other `if` before in the future.
```suggestion
            if (state[fieldSymbol] is { } value)
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs [hunk 6]
```diff
@@ -57,9 +57,18 @@ internal sealed class FieldReference : SimpleProcessor<IFieldReferenceOperationW
...
+            {
+                state = state.SetOperationValue(context.Operation, value);
+            }
+            if (fieldSymbol.Type.IsNonNullableValueType() || fieldSymbol.Type.IsEnum())
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:67]: This should have some UT in `RoslynSymbolicExecution.****.cs`

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:69]: I don't think this is enough. We have this problem in general for symbols. A similar change needs to be done also for parameters and local references.
So we'll need to extract it somewhere.
As such, `ISymbol.GetSymbolType()` extension can be used to get the type, instead of the `IFieldSymbol` check.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs:97]: We can do the change, especially when you have the context. But I'd split it in another PR as it has nothing to do with Enum.
Here, just accept the problem with ToDo

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:69]: The motivation for this is that the LocksReleasedAllPathsBase is doing `.SetSymbolConstraint(trackedCreatedNew, BoolConstraint.True),` without setting `NotNull`. Is that correct?

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
I'd prefer to split this into 3 PRs

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:69]: Is it really needed to fix the S2222 problem? It looks like a collateral finding.
